src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -474,120 +544,90 @@ namespace cryptonote
+    crypto::hash new_block_hash;
+    if (!parse_and_validate_block_from_blob(arg.b.block, new_block, &new_block_hash))
+    {
+      LOG_ERROR_CCONTEXT
+      (
+        "sent wrong block: failed to parse and validate block: "
+        << epee::string_tools::buff_to_hex_nodelimer(arg.b.block)
+        << ", dropping connection"
+      );
+
+      drop_connection(context, false, false);
+      return 1;
+    }
+
+    // Log block info
+    MLOG_P2P_MESSAGE(context << "Received NOTIFY_NEW_FLUFFY_BLOCK " << new_block_hash << " (height "
+      << arg.current_blockchain_height << ", " << arg.b.txs.size() << " txes)");
+
+    // If we are synchronizing the node or setting up this connection, then do nothing
     if(context.m_state != cryptonote_connection_context::state_normal)
       return 1;
     if(!is_synchronized()) // can happen if a peer connection goes to normal but another thread still hasn't finished adding queued blocks
     {
       LOG_DEBUG_CC(context, "Received new block while syncing, ignored");
       return 1;
     }
-    
+
+    // I don't know why we pause mining before we have finished verifying the block, seems like a
+    // DoS vector
     m_core.pause_mine();
-      
-    block new_block;
-    transaction miner_tx;
-    if(parse_and_validate_block_from_blob(arg.b.block, new_block))
+
     {
-      // This is a second notification, we must have asked for some missing tx
-      if(!context.m_requested_objects.empty())
+      // This set allows us to quickly sanity check that the block binds all txs contained in this
+      // fluffy payload, which means that no extra stowaway txs can be harbored. In the case of a
+      // deterministic block verification failure, the peer will be punished accordingly. For other
+      // cases, *once* valid PoW will be required to perform expensive concensus checks for the txs
+      // inside the block.
+      std::unordered_map<crypto::hash, uint64_t> blk_txids_set;
+      for (size_t tx_idx = 0; tx_idx < new_block.tx_hashes.size(); ++tx_idx)
+        blk_txids_set.emplace(new_block.tx_hashes[tx_idx], tx_idx);
+
+      // Check for duplicate txids in parsed block blob
+      if (blk_txids_set.size() != new_block.tx_hashes.size())
       {
-        // What we asked for != to what we received ..
-        if(context.m_requested_objects.size() != arg.b.txs.size())
-        {
-          LOG_ERROR_CCONTEXT
-          (
-            "NOTIFY_NEW_FLUFFY_BLOCK -> request/response mismatch, " 
-            << "block = " << epee::string_tools::pod_to_hex(get_blob_hash(arg.b.block))
-            << ", requested = " << context.m_requested_objects.size() 
-            << ", received = " << new_block.tx_hashes.size()
-            << ", dropping connection"
-          );
-          
-          drop_connection(context, false, false);
-          m_core.resume_mine();
-          return 1;
-        }
-      }      
+        MERROR("sent bad block entry: there are duplicate tx hashes in parsed block: "
```

Comment:
I _think_ you still need:
```c++
if (context.m_requested_objects.size() != arg.b.txs.size())
{ 
  // error
}
```
along with a check to ensure that every requested `txid` was provided by the peer.
?

Reply:
I'm like 80% sure that `m_requested_objects` is only relevant for the `NOTIFY_REQUEST_GET_OBJECTS`/`NOTIFY_RESPONSE_GET_OBJECTS` commands. When a node is missing txs in a fluffy block, they send a `NOTIFY_REQUEST_FLUFFY_MISSING_TX` command, which is responded to immediately with a new fluffy block containing the missing transactions. But these requests/responses don't touch `m_requested_objects`, which is for persistent, cross-command, state of block hashes.